blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClient.scala

```diff
@@ -63,69 +62,46 @@ object BlazeClient {
-                        .guarantee(manager.release(next.connection))
-                    case _ =>
-                      F.delay(stageOpt.foreach(_.removeStage()))
-                        .guarantee(manager.invalidate(next.connection))
-                  }
-                }
 
-              responseHeaderTimeout match {
-                case responseHeaderTimeout: FiniteDuration =>
-                  F.deferred[Unit].flatMap { gate =>
-                    val responseHeaderTimeoutF: F[TimeoutException] =
-                      F.delay {
-                        val stage =
-                          new ResponseHeaderTimeoutStage[ByteBuffer](
-                            responseHeaderTimeout,
-                            scheduler,
-                            ec)
-                        next.connection.spliceBefore(stage)
-                        stage
-                      }.bracket { stage =>
-                        F.async[TimeoutException] { cb =>
-                          F.delay(stage.init(cb)) >> gate.complete(()).as(None)
-                        }
-                      }(stage => F.delay(stage.removeStage()))
+            responseHeaderTimeout match {
+              case responseHeaderTimeout: FiniteDuration =>
+                Deferred[F, Unit].flatMap { gate =>
+                  val responseHeaderTimeoutF: F[TimeoutException] =
+                    F.delay {
+                      val stage =
+                        new ResponseHeaderTimeoutStage[ByteBuffer](
+                          responseHeaderTimeout,
+                          scheduler,
+                          ec)
+                      next.connection.spliceBefore(stage)
+                      stage
+                    }.bracket(stage =>
+                      F.async[TimeoutException] { cb =>
+                        F.delay(stage.init(cb)) >> gate.complete(()).as(None)
+                      })(stage => F.delay(stage.removeStage()))
 
-                    F.race(gate.get *> res, responseHeaderTimeoutF)
-                      .flatMap[Resource[F, Response[F]]] {
-                        case Left(r) => F.pure(r)
-                        case Right(t) => F.raiseError(t)
-                      }
-                  }
-                case _ => res
-              }
+                  F.racePair(gate.get *> res, responseHeaderTimeoutF)
+                    .flatMap[Resource[F, Response[F]]] {
+                      case Left((outcome, fiber)) =>
+                        fiber.cancel >> outcome.embed(
+                          F.raiseError(new CancellationException("Response canceled")))
+                      case Right((fiber, outcome)) =>
+                        fiber.cancel >> outcome.fold(
+                          F.raiseError(new TimeoutException("Response timeout also timed out")),
+                          F.raiseError,
+                          _.flatMap(F.raiseError)
+                        )
+                    }
+                }
```

Comment:
These `outcome` folds are novel work.  I think they're right, but perhaps not idiomatic.  The "Response timeout also timed out" situation would be particularly bizarre, but the types made me handle it.